syntax = "proto3";
package chat;

// Define the different types of actions.
enum Action {
  UNKNOWN = 0;
  CONNECT = 1;
  SEND_MESSAGE = 2;
  NOTIFY_ALL = 3;
  SEND_TO = 4;
}

message ChatRequest {
  Action action = 1;
  // For CONNECT, SEND_MESSAGE, and NOTIFY_ALL, this is the sender's username.
  string username = 2;
  // For SEND_TO, this is the intended recipient.
  string target = 3;
  // The content of the message.
  string message = 4;
}

message ChatResponse {
  bool success = 1;
  string message = 2;
}

service ChatService {
  // A single bidirectional stream for all chat operations.
  rpc Chat(stream ChatRequest) returns (stream ChatResponse);
}